README.md
# MagicMirrorModule-YouLess

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/). This module reads from a YouLess energy meter in your network to display realtime energy usage information.

This module is not yet finished; I intend to add more features and improve upon the display of the energy reading. (e.g. displaying power usage using an analog dial, sliding scale, history, graph, et cetera)

## Installation

1. Navigate into your MagicMirror's `modules` folder and run:
```
$ git clone https://github.com/QNimbus/MMM-YouLess.git
```
1. Install the dependencies: 
```
$ cd MMM-YouLess && npm install --only=production
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-YouLess',
            config: {
                // See below for configurable options
                youlessHost: "10.81.1.22", // or youless.home.network.nl
                updateInterval: 5000,
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `youlessHost`| *Required* Your YouLess ip address or hostname<br><br>**Type:** `string`
| `symbol`| *Optional* Font Awesome icon to display in the module<br><br>**Default value:** `bolt`<br>**Type:** `string`<br>See: [Font Awesome Icons website](http://fontawesome.io/icons/)
| `updateInterval` | *Optional* Interval at which the module fetches new data<br><br>**Default value:** 10000 milliseconds (10 seconds)<br>**Type:** `int`(milliseconds)<br>
| `initialLoadDelay` | *Optional* Initial delay when module is first loaded<br><br>**Default value:** 0 milliseconds (0 seconds)<br>**Type:** `int`(milliseconds)<br>
| `retryDelay`     | *Optional* If no data is received, retry again after delay<br><br>**Default value:** 2000 milliseconds (2 seconds)<br>**Type:** `int`(milliseconds)<br>

## Building from TypeScript source

Todo...

## Screenshots

![alt text][ss_01]

[ss_01]: images/module_ss_01.png "Example of YouLess module at work"

![alt text][ss_02]

[ss_02]: images/module_ss_02.png "Example of YouLess module at work"

CHANGELOG.md
#Changlelog

### [0.1.1] - Changed several instanced of bound functions to use the 'var self = this;' pattern since benchmarks show that it's faster that way. (https://jsperf.com/bind-vs-self-closure)

### [0.1.0] - Initial commit

node_helper.js
/* Magic Mirror
 * Node Helper: MMM-YouLess
 *
 * By B. van Wetten
 * MIT Licensed.
 */

var request = require("request");
var NodeHelper = require("node_helper");

var nodeHelper = NodeHelper.create({
	config: {},

	// Override start method.
	start: function () {
	},

	// Override socketNotificationReceived method.
	socketNotificationReceived: function (notification, payload) {
		var self = this;

		var config = function (payload) {
			self.config = payload;
		};

		var query = function (payload) {
			var options = {
				uri: `http://${self.config.youlessHost}/a?f=j`,
				method: "GET",
			};

			request(options, function (error, response, body) {
				if (response && response.statusCode === 200) {
					self.sendSocketNotification("RESPONSE", body);
				}
			});
		};

		var acceptedNotifications = {
			"CONFIG": config,
			"QUERY": query,
		};

		// Call appropriate handler for socket notification
		acceptedNotifications[notification] && typeof acceptedNotifications[notification] === "function" && acceptedNotifications[notification](payload);
	},
});

module.exports = nodeHelper;

MMM-YouLess.js
/* Magic Mirror
 * Module: MMM-YouLess
 *
 * By B. van Wetten
 * MIT Licensed.
 */

Module.register("MMM-YouLess", {
	// Module config defaults.
	defaults: {
		initialLoadDelay: 0, 		// 0 seconds delay
		retryDelay: 2 * 1000,		// retry after 2 seconds
		updateInterval: 10 * 1000, 	// every 10 seconds
		symbol: "bolt",
	},

	start: function () {
		this.loaded = false;
		this.powerUsage = -1;
		this.powerUsageHigh = -1;
		this.maxLevel = 1500;
		Log.info("Starting module: " + this.name);

		this.scheduleUpdate(this.config.initialLoadDelay);
	},

	getStyles: function () {
		return [
			this.file("css/youless.css"),
			"font-awesome.css",
		];
	},

	getScripts: function () {
		return [
			"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js",  // this file will be loaded from the jquery servers.
			this.file("jsGauge/lib/jsGauge.js"),
		];
	},

	getDom: function () {
		var wrapper = jQuery(document.createElement("div")).addClass("GaugeMeter").attr("id","YouLessGauge");;
		var obj = {
			background_style: "RGBa(255,255,255,0.6)",
			scale_style: "RGBa(255,255,255,0.6)",
			append: " kWh",
			size: 200,
			width: "15",
			style: "arc",
			theme: "Black",
			animate_gauge_colors: "1",
			animate_text_colors: "1",
			label: "Power usage",
			label_color: "#fff",
			stripe: "3",
			used: this.powerUsage,
			total: 5000,
		}
		wrapper.data(obj);
		wrapper.jsGauge();
		return wrapper[0];
	},

	updatePowerUsage: function () {
		var self = this;

		this.sendSocketNotification("QUERY");

		// Use default timeout if already loaded, else use the retryDelay if module hasn't loaded yet
		self.scheduleUpdate((self.loaded) ? -1 : self.config.retryDelay);
	},

	processPowerUsage: function (data) {
		if (data) {
			this.powerUsageHigh = this.powerUsageHigh < data.pwr ? data.pwr : this.powerUsageHigh;
			this.powerUsage = data.pwr;
			this.loaded = true;
			this.updateDom(this.config.animationSpeed);
		}
	},

	/* scheduleUpdate()
 	 * Schedule next update.
 	 *
 	 * argument delay number - Milliseconds before next update. If empty, this.config.updateInterval is used.
 	 */
	scheduleUpdate: function (delay) {
		var nextLoad = this.config.updateInterval;
		if (typeof delay !== "undefined" && delay >= 0) {
			nextLoad = delay;
		}

		var self = this;
		setTimeout(function () {
			self.updatePowerUsage();
		}, nextLoad);
	},

	notificationReceived: function (notification, payload, sender) {
		if (notification === "DOM_OBJECTS_CREATED") {
			this.loaded = true;
			this.sendSocketNotification("CONFIG", this.config);
		}
	},

	socketNotificationReceived: function (notification, payload) {
		var self = this;

		var responseReceived = function (payload) {
			self.processPowerUsage(JSON.parse(payload));
		};

		var acceptedNotifications = {
			"RESPONSE": responseReceived,
		};

		// Call appropriate handler for socket notification
		acceptedNotifications[notification] && typeof acceptedNotifications[notification] === "function" && acceptedNotifications[notification](payload);
	},
});